worker/raft/raftclusterer/worker_test.go

```diff
@@ -0,0 +1,302 @@
+			},
+			"2": {
+				ID:              "2",
+				InternalAddress: machine2Address,
+			},
+		},
+	})
+	waitConfiguration(c, s.Raft, []raft.Server{{
+		ID:       "machine-0",
+		Address:  raft.ServerAddress(machine0Address),
+		Suffrage: raft.Voter,
+	}, {
+		ID:       "machine-1",
+		Address:  raft.ServerAddress(machine1Address),
+		Suffrage: raft.Voter,
+	}, {
+		ID:       "machine-2",
+		Address:  raft.ServerAddress(machine2Address),
+		Suffrage: raft.Voter,
+	}})
+
+	select {
+	case <-raft1Observations:
+		c.Assert(raft1.Leader(), gc.Equals, s.Transport.LocalAddr())
+	case <-time.After(coretesting.LongWait):
+		c.Fatal("timed out waiting for leader observation")
+	}
+
+	// Remove machine-1, add machine-3.
+	s.publishDetails(c, apiserver.Details{
+		Servers: map[string]apiserver.APIServer{
+			"0": {
+				ID:              "0",
+				InternalAddress: machine0Address,
+			},
+			"2": {
+				ID:              "2",
+				InternalAddress: machine2Address,
+			},
+			"3": {
+				ID:              "3",
+				InternalAddress: machine3Address,
+			},
+		},
+	})
+	waitConfiguration(c, raft1, []raft.Server{{
+		ID:       "machine-0",
+		Address:  raft.ServerAddress(machine0Address),
+		Suffrage: raft.Voter,
+	}, {
+		ID:       "machine-2",
+		Address:  raft.ServerAddress(machine2Address),
+		Suffrage: raft.Voter,
+	}, {
+		ID:       "machine-3",
+		Address:  raft.ServerAddress(machine3Address),
+		Suffrage: raft.Voter,
+	}})
+}
+
+func (s *WorkerSuite) TestChangeLocalServer(c *gc.C) {
+	// Create 3 servers, machine-0, machine-1, and machine-2.
+	// The latter two servers can communicate bidirectionally;
+	// only machine-0 can dial the others, and not the other
+	// way around.
+	raft1, _, transport1, _, _ := s.NewRaft(c, "machine-1", &rafttest.FSM{})
+	_, _, transport2, _, _ := s.NewRaft(c, "machine-2", &rafttest.FSM{})
+	s.Transport.Connect(transport1.LocalAddr(), transport1)
+	s.Transport.Connect(transport2.LocalAddr(), transport2)
+	transport1.Connect(transport2.LocalAddr(), transport2)
+	transport2.Connect(transport1.LocalAddr(), transport1)
+	machine1Address := string(transport1.LocalAddr())
+	machine2Address := string(transport2.LocalAddr())
+
+	raft1Observations := make(chan raft.Observation, 10)
+	raft1Observer := raft.NewObserver(raft1Observations, false, func(o *raft.Observation) bool {
+		_, ok := o.Data.(raft.LeaderObservation)
+		return ok
+	})
+	raft1.RegisterObserver(raft1Observer)
+	defer close(raft1Observations)
+	defer raft1.DeregisterObserver(raft1Observer)
+
+	newLeaderElected := make(chan struct{})
+	go func() {
+		for {
+			select {
+			case _, ok := <-raft1Observations:
+				if !ok {
+					return
+				}
+			}
+			leaderAddr := raft1.Leader()
+			if leaderAddr != "" && leaderAddr != s.Transport.LocalAddr() {
+				close(newLeaderElected)
+				return
+			}
+		}
+	}()
+
+	// Here we simulate "ensure-ha": machine-0's address will
+	// be updated to a non-localhost address, and two new
+	// servers are added.
```

Comment:
I think this code can be updated, since we found that you *can* update existing server IP addresses.

Reply:
done